### DISCUSSION TITLE
Cheating prevention

### DISCUSSION THREAD

**Discussion by Vtec234** [piqueserver/piqueserver#639] [Feature Requests]
Currently the game is being ruined completely by the sheer number of cheaters present on every server. Votekicking exists, but people tend not to vote (and for good reason if they don't have proof of cheating), and server admins cannot be expected to spend much time on verifying reports and banning people.

I am not an AOS client or server developer, and so I don't know how the game is implemented or whether any of these suggested prevention measures have been implemented, so please bear with me.

Because players are in control of the clients, this has to be done on the server. Some cheating prevention can be implemented without breaking the network protocol, and so maintaining backwards compatibility. What could be done:
- To prevent wallhacks, the server could only send information about what each player can actually see to the players. This means some method of checking that on the server. A raycast through the world geometry from each player to each other player (n^2 operations) would be enough. There would also need to be heuristics so that players don't get information too late due to latency, so if the raycast "barely fails" (meaning the player could see the other in the next few frames), the information should be sent.
- To prevent aimbots, the server could check (especially on headshot kills) how quickly the player turned, so to speak. That means for example, how long did it take them since they observed the enemy (raycast above succeeded) to getting the headshot. If the numbers are consistently low, chances are the player's camera is snapping to enemies' heads quickly, which indicates an aimbot. Alternatively, just check the rotational speed of players' orientations if such a variable exists.

**Comment by NotAFile** [comment]
Thanks for submitting!

piqueserver already has server side aimbot detection (see: plugins/aimbot2.py) but many servers don't run the script and since the default is to just warn the admins, it generally just gets ignored.

**Comment by Vtec234** [comment]
Looking at the [config](https://github.com/piqueserver/piqueserver/blob/master/piqueserver/config/config.json) it looks disabled by default, so maybe a good first step would be to enable it. Then, it would be useful to, besides warning admins by default, also emit a server message like "aimbot detected" so that people have reason to votekick the cheater.

**Comment by noway** [comment]
That can be configured with aimbot indeed. 

Re: raicasting, it's not really possible with the way aos protocol is designed AFAIK.

**Comment by xtreme8000** [comment]
> Re: raicasting, it's not really possible with the way aos protocol is designed AFAIK

It is possible the way the protocol is laid out, but it will certainly make servers lag a lot more to check for block intersections for every player on every other player and every tick.

**Comment by NotAFile** [comment]
I'm not sure how reliable or accurate the aimbot2 script is, so the reason it's disabled might be that it's just not very good. I'm not sure how we would test that though seeing as none of us has a popular server.

**Comment by noway** [comment]
Right, now i understand what you mean. Yep, raycasting as in hiding not visible players from network packets is an interesting idea which could be implemnted with legacy clients

**Comment by NotAFile** [comment]
I am worried about the computational overhead of the raycasting method too.

You'd have to cast 0-30 rays 30-60 times a second for 1-30 players — that's a lot of rays

**Comment by yvt** [comment]
Exactly speaking, simple ray casting is not sufficient. Imagine a case where you only can see a part of another player. However, solving an exact solution requires high-dimensional CSG, which is infeasible in every way I can think of. 

Ray casting using an eroded (by 1 or 2 blocks, depending on ping and movement speeds) map geometry might provide a good conservative approximation to this. Mipmapping, stochastic sampling, and acceleration structures are also good for performance.

By the way, sound is a very important factor in this game. Even if you can't see them directly, you can spot the hidden foes by noises they are making. Because of this, sending information about invisible players is still necessary in some cases.

**Comment by NotAFile** [comment]
Oh yeah, the sound thing is a very good point... Even if we didn't send players, we would still have to send sound, which requires sending a position. 

Erosion sounds pretty great, I'm not sure what the other two are but I'll look it up.
